README.md
```mermaid
graph TB
    main["main.go"] -->|coordinates| USBDeviceDelegate
    main -->|handles callbacks| ConnectedDevices
    main -->|manages connection| TCPSocket
    USBDeviceDelegate -->|listens for events| USB
    USBDeviceDelegate -->|manages active connection| ConnectedDevices
    USBDeviceDelegate -->|opens connection| TCPSocket
    USB -->|handles connection| Scan
    USB -->|handles data transfer| ConnectedDevices
    USB -->|handles event callbacks| CallbackInterfaces
    Scan -->|checks for interrupts| USBDevice
    USBDevice -->|handles callbacks| USBDeviceDelegate
    ConnectedDevices -->|handles data transfer| USBDevice
    ConnectedDevices -->|handles event callbacks| ConnectedDeviceDelegate
    TCPSocket -->|handles TCP tunnel connection| usbmuxd
    CallbackInterfaces -->|callback events| USB
    CallbackInterfaces -->|callback events| ConnectedDevices
    ConnectedDeviceDelegate -->|handles event callbacks| ConnectedDevices
    ConnectedDevices -->|handles data transfer| FrameEncoding
    FrameEncoding -->|translates to binary| ProtocolFrames
    FrameEncoding -->|translates to dictionary| PLISTEncoding
    ProtocolFrames -->|handles requests and responses| FrameEncoding
    PLISTEncoding -->|handles data compacting| FrameEncoding
```

# Overview

This project enables communicating with an iOS device like an iPhone or iPad over USB using the `usbmuxd` daemon on the device. It allows detecting device connections/disconnections, connecting to a TCP port on the device, and sending/receiving data.

The key components are:

* USB device scanning and connection handling
* TCP socket communication
* Parsing protocol frames for sending requests and handling responses
* Callback interfaces for event handling

## Main Packages

**main.go**: The main module coordinates scanning for devices, connecting to them, and handling callbacks.

* The `USBDeviceDelegate` struct implements callback methods for USB events.
* `USB.Listen()` is used to listen for device attach/detach events.
* `USB.ConnectedDevices` manages the active USB connection.
* A TCP socket connection is opened to `usbmuxd` for communication.

**USB/** - Contains code for USB device management

* `Scan` periodically tries connecting to USB devices until successful.
* `ConnectedDevices` represents an active USB connection for sending/receiving data.
* Callback interfaces like `ConnectedDeviceDelegate` handle events.

**frames/** - Defines structures for protocol frames and PLIST encoding

* USB request frames like `USBListenRequestFrame`
* Response frames like `USBGenericACKFrame`
* Attach/detach notice frames like `USBDeviceAttachedDetachedFrame`

**transmission/** - Manages the TCP tunnel connection to `usbmuxd`

## Device Scanning

The `Scan` struct periodically tries connecting to a USB device whenever it is plugged in. It runs in a goroutine checking for interrupts.

When a device is attached, the `USBDeviceDidPlug` callback is invoked, which starts a scan. The scan tries connecting every second until successful.

On success, the `USBDeviceDidSuccessfullyConnect` callback is called. On failure, `USBDeviceDidFailToConnect` is called, and scanning is restarted.

## Sending and Receiving Data

Once connected, the `ConnectedDevices` instance can be used to send and receive data.

Data is sent by constructing a header with metadata like the message tag, followed by the payload data. This is then written to the USB TCP socket.

Received data is parsed by checking the header fields, extracting the tag and payload, and passing it to the `USBDeviceDidReceiveData` callback.

The connection is closed when the socket returns EOF, invoking the `USBDeviceDidDisconnect` callback to restart scanning.

## Frame Encoding

Frames are encoded using the PLIST binary format. This allows sending dictionary structures over the protocol while keeping them compact.

The frame types define the expected keys and data types used in requests and responses. They are serialized to binary when sending and deserialized when receiving.

## Event Handling

The `USBDeviceDelegate` implements callback methods for various USB events and data. These handle the core logic in response to device lifecycle and I/O.

Other delegates like `ConnectedDeviceDelegate` handle events related to an active USB connection.

Callbacks are invoked by the `USB` and `ConnectedDevices` instances to notify consumers of state changes and data. This separation of concerns keeps the code modular and extensible.

## Conclusion

This covers the major components and flow for communicating with iOS devices over USB in Go using the `usbmuxd` protocol. The project shows how to integrate device management, network I/O, protocol encoding, and concurrent event handling to build a robust USB client.


CHANGES.md
# Audio Streaming Between an iPhone and a Browser JavaScript App via a VPS

To enable audio streaming between an iPhone and a browser JavaScript app via a VPS, we need to modify the Go code. Below are the necessary steps:

## 1. Update the TCP Socket Connection

Update the TCP socket connection code to connect to the VPS IP and port instead of localhost.

```go
// File: transmission/tunnel.go

var tcpAddress = "VPS_IP:VPS_PORT"

func Tunnel() net.Conn {
  conn, err := net.Dial(tcpPort, tcpAddress)
  // Handle error if any
}
```

## 2. Modify Main Code to Listen for Audio

In the main code, listen on a local port for audio sent from the JavaScript app. Pass the received data to the USB send code.

```go
// File: main.go

ln, _ := net.Listen("tcp", ":LOCAL_PORT")
go func() {
  for {
    conn, _ := ln.Accept()
    // Get audio data from conn
    // Pass to usb.SendData()
  }
}()
```

## 3. Forward Audio Data from iPhone to VPS

In the USB receive callback, forward the audio data received from iPhone to the VPS port.

```go
// File: USB/connected_devices.go

func (d ConnectedDeviceDelegate) USBDeviceDidReceiveData(data []byte) {

  conn, _ := net.Dial("tcp", "VPS_IP:VPS_PORT")

  // Send data to VPS
}
```

## 4. Audio Input via JavaScript

In JavaScript, obtain the audio input using the Web Audio API and send it to the Go code listening on the VPS port.

```javascript
// Get input stream from microphone

const source = audioContext.createMediaStreamSource(stream)

// Send audio data to VPS port
const socket = new WebSocket('ws://VPS_IP:VPS_PORT') 
source.connect(socket)
```

## 5. Receive and Play Audio Data

Finally, receive the audio data on the VPS port and play it using the Web Audio API.

```javascript
// Get audio stream from VPS port

const socket = new WebSocket('ws://VPS_IP:VPS_PORT')

// Play received audio
socket.onmessage = function(e) {
  audioContext.decodeAudioData(e.data, function(buffer) {
    const source = audioContext.createBufferSource()
    source.buffer = buffer
    source.connect(audioContext.destination)
    source.start()
  })
}
```


main.go
package main

import (
	"context"
	"io"
	"log"
	"net"
	"os"
	"os/signal"
	"syscall"
	"time"

	"go-usbmuxd/USB"
	"go-usbmuxd/frames"
	"go-usbmuxd/transmission"

	"nhooyr.io/websocket"
)

// some global vars
var connectHandle USB.ConnectedDevices
var port = 29173
var pluggedUSBDevices map[int]frames.USBDeviceAttachedDetachedFrame
var connectedUSB int // only stores the device id
var scanningInstance USB.Scan
var self USBDeviceDelegate

var websocketProxyHost string = "ws://49.13.56.241:6969/usbmux/client/subscribe"

func main() {
	// inti section
	connectedUSB = -1
	pluggedUSBDevices = map[int]frames.USBDeviceAttachedDetachedFrame{}
	scanningInstance = USB.Scan{}
	self = USBDeviceDelegate{}

	// logger
	logFile, err := os.OpenFile("kusb_ios.log", os.O_CREATE|os.O_APPEND|os.O_RDWR, 0666)
	if err != nil {
		log.Println(err)
	}
	defer logFile.Close()
	mw := io.MultiWriter(os.Stdout, logFile)
	log.SetOutput(mw)

	var (
		listenConnection net.Conn
	)
	defer func() {
		if r := recover(); r != nil {
			listenConnection.Close()
			connectHandle.Connection.Close()
			scanningInstance.Stop()
			log.Printf("Error %v", r)
		}
	}()

	// create a USB.Listen(USBDeviceDelegate) instance. Pass a delegate to resolve the attached and detached callbacks
	// then on device added save ot to array/ map and send connect to a port with proper tag
	listenConnection = USB.Listen(transmission.Tunnel(), self)
	// defer listenConnection.Close()

	// connect to a random usb device, if Number == 0 then
	connectHandle = USB.ConnectedDevices{Delegate: self, Connection: transmission.Tunnel()}
	// defer connectHandle.Connection.Close()

	// scan defer
	// defer scanningInstance.Stop()

	// go func() {
	// 	for {
	// 		if len(pluggedUSBDevices) != 0 {
	// 			log.Printf("DEBUG %v", pluggedUSBDevices)
	// 		}
	// 	}
	// }()

	go func() {
		ctx, cancel := context.WithTimeout(context.TODO(), time.Second*10)
		defer cancel()
		conn, _, err := websocket.Dial(ctx, websocketProxyHost, nil)
		if err != nil {
			log.Printf("Error: %v", err)
			return
		}
		defer conn.Close(websocket.StatusNormalClosure, "")
		for {
			_, message, err := conn.Read(context.TODO())
			if err != nil {
				log.Printf("Read Error: %v", err)
				return
			}
			connectHandle.SendData(message, 106)
		}
	}()

	quit := make(chan os.Signal)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGKILL, syscall.SIGTERM)

	// run loop
	select {
	case <-quit:
		listenConnection.Close()
		connectHandle.Connection.Close()
		scanningInstance.Stop()
	}
}

// USBDeviceDelegate - USB Delegate Methods
type USBDeviceDelegate struct{}

// USBDeviceDidPlug - device plugged callback
func (usb USBDeviceDelegate) USBDeviceDidPlug(frame frames.USBDeviceAttachedDetachedFrame) {
	// usb has been plugged DO: startScanning
	log.Printf("[USB-INFO] : Device Plugged %s ID: %d\n", frame.Properties.SerialNumber, frame.DeviceID)
	pluggedUSBDevices[frame.DeviceID] = frame
	scanningInstance.Start(&connectHandle, frame, port)
	connectHandle.SendData([]byte("audio data"), 1)
}

// USBDeviceDidUnPlug - device unplugged callback
func (usb USBDeviceDelegate) USBDeviceDidUnPlug(frame frames.USBDeviceAttachedDetachedFrame) {
	// usb has been unplugged
	// stop scan
	log.Printf("[USB-INFO] : Device UnPlugged %s ID: %d\n", pluggedUSBDevices[frame.DeviceID].Properties.SerialNumber, frame.DeviceID)
	delete(pluggedUSBDevices, frame.DeviceID)
	scanningInstance.Stop()
}

// USBDidReceiveErrorWhilePluggingOrUnplugging - device plugging/unplugging callback
func (usb USBDeviceDelegate) USBDidReceiveErrorWhilePluggingOrUnplugging(err error, stringResponse string) {
	// plug or unplug error
	// stop scan
	if stringResponse != "" {
		//some unresolved message came
		//TODO - Implement some resolver to understand message received
	}
	log.Println("[USB-EM-1] : Some error encountered wile pluging and unpluging. ", err.Error())
	scanningInstance.Stop()
}

// USBDeviceDidSuccessfullyConnect - device successful connection callback
func (usb USBDeviceDelegate) USBDeviceDidSuccessfullyConnect(device USB.ConnectedDevices, deviceID int, toPort int) {
	// successfully connected to the port mentioned
	// stop the scan
	connectedUSB = deviceID
	log.Printf("HERE STOP SCAN")
	scanningInstance.Stop()
}

// USBDeviceDidFailToConnect - device connection failure callback
func (usb USBDeviceDelegate) USBDeviceDidFailToConnect(device USB.ConnectedDevices, deviceID int, toPort int, err error) {
	// error while communication in the socket
	// start scan
	connectedUSB = -1
	pluggedDeviceID := getFirstPluggedDeviceId()
	if pluggedDeviceID != -1 {
		scanningInstance.Start(&connectHandle, pluggedUSBDevices[pluggedDeviceID], port)
	}

}

// USBDeviceDidReceiveData - data received callback
func (usb USBDeviceDelegate) USBDeviceDidReceiveData(device USB.ConnectedDevices, deviceID int, messageTAG uint32, data []byte) {
	// received data from the device
	log.Println(string(data))
	//device.SendData(data[20:], 106)
}

// USBDeviceDidDisconnect - device disconnect callback
func (usb USBDeviceDelegate) USBDeviceDidDisconnect(devices USB.ConnectedDevices, deviceID int, toPort int) {
	// socket disconnect
	// start scan
	connectedUSB = -1
	pluggedDeviceID := getFirstPluggedDeviceId()
	if pluggedDeviceID != -1 {
		scanningInstance.Start(&connectHandle, pluggedUSBDevices[pluggedDeviceID], port)
	}
}

// MARK - helper functions here
// Needs restructuring, removal or other implementation
func getFirstPluggedDeviceId() int {
	var deviceID int = -1
	for deviceID, _ = range pluggedUSBDevices {
		break
	}
	return deviceID
}


USB/scan.go
package USB

import (
	"fmt"
	"time"

	"go-usbmuxd/frames"
	"go-usbmuxd/transmission"
)

type (
	Scan struct {
		IsScanning bool
	}
)

func (scan *Scan) Start(device *ConnectedDevices, frame frames.USBDeviceAttachedDetachedFrame, toPort int) {
	if false == scan.IsScanning {
		// set scanning to true and scan
		scan.IsScanning = true
		// scanning routine
		go func() {
			// scan here
			// first check the chan for any interrupt
			// if interrupt is there ... break the loop and set isScanning false
			for i := 0; i < 300; i++ {
				if scan.IsScanning && device.Connection != nil {
					device.Connect(transmission.Tunnel(), frame, toPort)
				} else {
					fmt.Println("BREAK CONNECT")
					break
				}
				time.Sleep(time.Second * 1)
			}
			if scan.IsScanning {
				scan.IsScanning = false
			}
		}()
	}
}
func (scan *Scan) Stop() {
	if true == scan.IsScanning {
		scan.IsScanning = false
	}
}


USB/connect.go
package USB

import (
	"bytes"
	"encoding/binary"
	"errors"
	"io"
	"log"
	"net"

	"howett.net/plist"

	"go-usbmuxd/frames"
)

type (
	ConnectedDeviceDelegate interface {
		USBDeviceDidSuccessfullyConnect(device ConnectedDevices, deviceID int, toPort int)
		USBDeviceDidFailToConnect(device ConnectedDevices, deviceID int, toPort int, err error)
		USBDeviceDidReceiveData(device ConnectedDevices, deviceID int, messageTAG uint32, data []byte)
		USBDeviceDidDisconnect(devices ConnectedDevices, deviceID int, toPort int)
	}
	ConnectedDevices struct {
		Delegate   ConnectedDeviceDelegate
		Connection net.Conn
	}
)

func (device ConnectedDevices) Connect(conn net.Conn, frame frames.USBDeviceAttachedDetachedFrame, port int) net.Conn {
	device.Connection.Close()
	device.Connection = conn
	go connectFrameParser(device.Connection, frame.DeviceID, port, device)

	// now send a connect request to the device
	device.Connection.Write(sendConnectRequestToSocket(frame.DeviceID, port))

	return device.Connection
}

func (device ConnectedDevices) SendData(data []byte, messageTagType uint32) {
	// create a 20byte standard header that's used by peertalk to parse tag and other info
	headerBuffer := make([]byte, 20)

	// preparing the header
	binary.BigEndian.PutUint32(headerBuffer[:4], 1)
	binary.BigEndian.PutUint32(headerBuffer[4:8], messageTagType)
	binary.BigEndian.PutUint32(headerBuffer[8:12], 0)
	binary.BigEndian.PutUint32(headerBuffer[12:16], uint32(len(data)+4))
	binary.BigEndian.PutUint32(headerBuffer[16:], uint32(len(data)))

	if device.Connection != nil {
		_, err := device.Connection.Write(append(headerBuffer, data...))
		if err != nil {
			log.Println(err.Error())
		}
	}

	log.Printf("SENT DATA %v", string(data))
}

func byteSwap(val int) int {
	return ((val & 0xFF) << 8) | ((val >> 8) & 0xFF)
}

func sendConnectRequestToSocket(deviceID int, toPort int) []byte {
	// constructing body
	body := &frames.USBConnectRequestFrame{
		DeviceID:            deviceID,
		PortNumber:          byteSwap(toPort),
		MessageType:         "Connect",
		ClientVersionString: "1.0.0",
		ProgName:            "go-usbmuxd",
	}
	bodyBuffer := &bytes.Buffer{}
	encoder := plist.NewEncoder(bodyBuffer)
	encoder.Encode(body)

	//constructing header
	headerBuffer := make([]byte, 16)
	binary.LittleEndian.PutUint32(headerBuffer, uint32(bodyBuffer.Len()+16))
	headerBuffer[4] = byte(1)
	headerBuffer[8] = byte(8)
	headerBuffer[12] = byte(1)

	requestBuffer := append(headerBuffer, bodyBuffer.Bytes()...)
	return requestBuffer
}

func connectFrameParser(conn net.Conn, deviceID int, toPort int, device ConnectedDevices) {
	chunk := make([]byte, 2500000)
	for {
		n, err := conn.Read(chunk)
		if err != nil {
			if err != io.EOF {
				panic("[USB-ERROR-iCONNECT-1] : Unable to read data stream from the USB channel")
			}
			device.Delegate.USBDeviceDidDisconnect(device, deviceID, toPort)
			break
		}
		// initial check for message type
		var data frames.USBGenericACKFrame
		decoder := plist.NewDecoder(bytes.NewReader(chunk[16:n]))
		decoder.Decode(&data)
		if data.MessageType == "Result" && data.Number == 0 {
			device.Delegate.USBDeviceDidSuccessfullyConnect(device, deviceID, toPort)
		} else if data.MessageType == "Result" && data.Number != 0 {
			var errorMessage string
			switch data.Number {
			case 2:
				// Device Disconnected
				errorMessage = "Unable to connect to Device, might be issue with the cable or turned off"
			case 3:
				// Port isn't available/ busy
				errorMessage = "Port you're requesting is unavailable"
			case 5:
				// UNKNOWN Error
				errorMessage = "[IDK]: Malformed request received in the device"
			}
			device.Delegate.USBDeviceDidFailToConnect(device, deviceID, toPort, errors.New(errorMessage))
		}
		if data.MessageType != "Result" {
			// parse the TAG and other relevant header info
			headerBuffer := chunk[:16]
			//log.Println(binary.BigEndian.Uint32(headerBuffer[0:4]))
			//log.Println(binary.BigEndian.Uint32(headerBuffer[4:8]))
			//log.Println(binary.BigEndian.Uint32(headerBuffer[8:12]))
			//log.Println(binary.BigEndian.Uint32(headerBuffer[12:16]))
			//log.Println(binary.BigEndian.Uint32(headerBuffer[16:20]))
			device.Delegate.USBDeviceDidReceiveData(device, deviceID, binary.BigEndian.Uint32(headerBuffer[4:8]), chunk[:n])
		}
	}
}


USB/listen.go
package USB

import (
	"bytes"
	"encoding/binary"
	"errors"
	"net"

	"go-usbmuxd/frames"

	"howett.net/plist"
)

func Listen(conn net.Conn, delegate USBDeviceDelegate) net.Conn {
	// start a tunnel here, and then send the listen frame to that connected socket
	go frameParser(conn, delegate)

	// send a listen request to usbmuxd daemon socket
	conn.Write(sendListenRequestToSocket())

	return conn
}

func frameParser(conn net.Conn, delegate USBDeviceDelegate) {
	chunk := make([]byte, 2500000)
	for {
		n, err := conn.Read(chunk)
		if err != nil {
			panic("[USB-ERROR-READ-1] : Unable to read data stream from the USB channel")
		}
		// initial check for message type
		var data frames.USBGenericACKFrame
		decoder := plist.NewDecoder(bytes.NewReader(chunk[16:n]))
		decoder.Decode(&data)

		if data.MessageType == "Result" && data.Number != 0 {
			delegate.USBDidReceiveErrorWhilePluggingOrUnplugging(errors.New("[USB-ERROR-LISTEN-RESP-1] : Illegal response received"), string(chunk[16:n]))
		}
		if data.MessageType != "Result" {
			var data frames.USBDeviceAttachedDetachedFrame
			decoder = plist.NewDecoder(bytes.NewReader(chunk[16:n]))
			decoder.Decode(&data)
			if data.MessageType == "Attached" {
				delegate.USBDeviceDidPlug(data)
			} else if data.MessageType == "Detached" {
				delegate.USBDeviceDidUnPlug(data)
			} else {
				delegate.USBDidReceiveErrorWhilePluggingOrUnplugging(errors.New("[USB-ERROR-LISTEN-2] : Unable to parse the response"), string(chunk[16:n]))
			}
		}
	}
}

func sendListenRequestToSocket() []byte {
	// constructing body
	body := &frames.USBListenRequestFrame{
		MessageType:         "Listen",
		ProgName:            "go-usbmuxd",
		ClientVersionString: "1.0.0",
	}
	bodyBuffer := &bytes.Buffer{}
	encoder := plist.NewEncoder(bodyBuffer)
	encoder.Encode(body)

	//constructing header
	headerBuffer := make([]byte, 16)
	binary.LittleEndian.PutUint32(headerBuffer, uint32(bodyBuffer.Len()+16))
	headerBuffer[4] = byte(1)
	headerBuffer[8] = byte(8)
	headerBuffer[12] = byte(1)

	requestBuffer := append(headerBuffer, bodyBuffer.Bytes()...)
	return requestBuffer
}
